src/main/java/redis/clients/jedis/HostAndPort.java

```diff
@@ -90,15 +90,50 @@ public static HostAndPort parseString(String from){
   }
 
   public static String convertHost(String host) {
-    if (host.equals("127.0.0.1") || host.startsWith("localhost") || host.equals("0.0.0.0") ||
-        host.startsWith("169.254") ||
-        host.startsWith("::1") || host.startsWith("0:0:0:0:0:0:0:1")) {
-      return LOCALHOST_STR;
-    } else {
+    try {
+        /*
+         * Validate the host name as an IPV4/IPV6 address.
+         * If this is an AWS ENDPOINT it will not parse.
+         * In that case accept host as is.
+         *
+         * Costs: If this is an IPV4/6 encoding, e.g. 127.0.0.1 then no DNS lookup
+         * is done.  If it is a name then a DNS lookup is done but it is normally cached.
+         * Secondarily, this class is typically used to create a connection once
+         * at the beginning of processing and then not used again.  So even if the DNS
+         * lookup needs to be done then the cost is miniscule.
+         */
+      InetAddress inetAddress = InetAddress.getByName(host);
+
+      // isLoopbackAddress() handles both IPV4 and IPV6
+      if (inetAddress.isLoopbackAddress() || host.equals("0.0.0.0") || host.startsWith("169.254"))
+        return getLocalhost();
+      else
+        return host;
+    } catch (Exception e) {
+      // Not a valid IP address
+      log.logp(Level.WARNING, HostAndPort.class.getName(), "convertHost", "'" + host + "' is not a valid IP address",
+              e);
       return host;
     }
   }
 
+  public static void setLocalhost(String localhost) {
```

Comment:
Are there any actual usages for `setLocalhost()`? `setLocalhost()` and `getLocalhost()` can make race condition. 

Btw, `getLocalhost()` itself can make race condition but it might be OK given that `getLocalHostQuietly()` always returns same value.